tests/gtest_general_status_control.cpp

```diff
@@ -0,0 +1,350 @@
+    }
+    const auto& status = root_->getGeneralStatus();
+    ASSERT_TRUE(status.has_value()) << "Test " << name << " final status not set";
+    EXPECT_EQ(status.value().status_code_, expectedCode)
+        << "Test " << name << " final code mismatch";
+  }
+
+  void resetRoot()
+  {
+    root_->halt();   // this should be already called, but it is still sometimes needed
+    root_->resetGeneralStatus();
+  }
+};
+
+/****************TESTS START HERE***************************/
+TEST_F(GeneralStatusControlTest, Fallback)
+{
+  // Fallback
+  root_ = std::make_unique<BT::FallbackNode>("FallbackNode");
+  root_->addChild(actions_[0].get());
+  root_->addChild(actions_[1].get());
+
+  // 1: A(FAIL) -> B(OK) -> OUTPUT: OK
+  setExpectedResults({NodeStatus::FAILURE, NodeStatus::SUCCESS});
+  runTest({NodeStatus::RUNNING, NodeStatus::RUNNING, NodeStatus::SUCCESS},
+          BtErrorCodes::OK, "1");
+
+  // 2: A(FAIL) -> B(FAIL) -> OUTPUT: FAIL (B)
+  setExpectedResults({NodeStatus::FAILURE, NodeStatus::FAILURE});
+  runTest({NodeStatus::RUNNING, NodeStatus::RUNNING, NodeStatus::FAILURE},
+          getExpectedCode(1), "2");
+
+  SUCCEED();
+}
+
+TEST_F(GeneralStatusControlTest, IfThenElse)
+{
+  // [BUG]: IfThenElseNode is not halting child A if there is no child C
+
+  // IfThenElse (A&B only)
+  root_ = std::make_unique<BT::IfThenElseNode>("IfThenElseNode");
+  root_->addChild(actions_[0].get());
+  root_->addChild(actions_[1].get());
+
+  // 1(no C): A(FAIL) -> FAIL (generic error)
+  setExpectedResults({NodeStatus::FAILURE, NodeStatus::FAILURE});
+  runTest({NodeStatus::RUNNING, NodeStatus::FAILURE},
+          BtErrorCodes::BEHAVIOR_TREE_NODE_FAILURE, "1");
+
+  // Add third child (C)
+  root_->addChild(actions_[2].get());
+
+  // 2: A(FAIL) -> C(FAIL) -> FAIL
+  setExpectedResults({NodeStatus::FAILURE, NodeStatus::FAILURE, NodeStatus::FAILURE});
+  runTest({NodeStatus::RUNNING, NodeStatus::RUNNING, NodeStatus::FAILURE},
+          getExpectedCode(2), "2");
+
+  // 3: A(FAIL) -> C(OK) -> OK
+  setExpectedResults({NodeStatus::FAILURE, NodeStatus::FAILURE, NodeStatus::SUCCESS});
+  runTest({NodeStatus::RUNNING, NodeStatus::RUNNING, NodeStatus::SUCCESS},
+          BtErrorCodes::OK, "3");
+
+  // 4: A(OK) -> B(FAIL) -> FAIL
+  setExpectedResults({NodeStatus::SUCCESS, NodeStatus::FAILURE, NodeStatus::SUCCESS});
+  runTest({NodeStatus::RUNNING, NodeStatus::RUNNING, NodeStatus::FAILURE},
+          getExpectedCode(1), "4");
+
+  // 5: A(OK) -> B(OK) -> OK
+  setExpectedResults({NodeStatus::SUCCESS, NodeStatus::SUCCESS, NodeStatus::SUCCESS});
+  runTest({NodeStatus::RUNNING, NodeStatus::RUNNING, NodeStatus::SUCCESS},
+          BtErrorCodes::OK, "5");
+
+  SUCCEED();
+}
+
+// Manual node is not trivial to test automatically
+// TEST_F(GeneralStatusControlTest, Manual)
+// {
+//   SUCCEED();
+// }
+
+TEST_F(GeneralStatusControlTest, Parallel)
+{
+  // Parallel
+  // There is no propagation in parallel node, because it can have multiple failed children
```

Comment:
This is something that will definitely hurt us. We use parallel for e.g. continuous reflector finepos and need to report a proper error here.

I don't immediately have a solution in mind, but we'll have to think about some way of moving the burden of propagation to the user/BTree if we can't decide on our own.

Reply:
BT library do not have capability to deduct, which error should be propagated if two children fail in the parallel node.

However, we can assume, that children in ParallelNode have decreasing priority: 1>2>3>.... Then if, for example, children no 2 and 3 would fail, the node will propagate error from children 2 to upper layer. 

Then for most of the cases, correct structure of tree would allow propagation of most significant children error. For more specialized case, the specialized parallel control node would be needed.

Reply:
After discussion with @gstrommer I added a specialized error code in case of multiple failed nodes. In most cases, one failure is enough to stop Parallel Node, so it will propagate the failed child error code.